external/llvm-project/mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp

```diff
@@ -830,6 +830,151 @@ LogicalResult PackedStochRoundFp8OpLowering::matchAndRewrite(
   return success();
 }
 
+// Implement the AMDGPU_DPPLowering class that will convert the amdgpu.dpp
+// operation into the corresponding ROCDL instructions.
+struct AMDGPUDPPLowering : public ConvertOpToLLVMPattern<DPPOp> {
+  AMDGPUDPPLowering(LLVMTypeConverter &converter, Chipset chipset)
+      : ConvertOpToLLVMPattern<DPPOp>(converter), chipset(chipset) {}
+  Chipset chipset;
+
+  Value createConstantOrUseDefault(Operation *op,
+                                   ConversionPatternRewriter &rewriter,
+                                   StringRef attrName, Type type) const {
+    auto attr = op->getAttr(attrName);
+    return rewriter.create<LLVM::ConstantOp>(op->getLoc(), type, attr);
+  }
+
+  LogicalResult
+  matchAndRewrite(DPPOp DppOp, DPPOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    // Convert the source operand to the corresponding LLVM type
+    Location loc = DppOp.getLoc();
+    Value src = adaptor.getSrc();
+    Type srcType = src.getType();
+    auto llvmI32Type = typeConverter->convertType(rewriter.getI32Type());
+    auto llvmI1Type = typeConverter->convertType(rewriter.getI1Type());
+    auto llvmSrcIntType = typeConverter->convertType(
+        rewriter.getIntegerType(srcType.getIntOrFloatBitWidth()));
+    Value dppCtrl;
+
+    // If the source type is less or equal to i32 or f32, use bitcast to convert
+    // it to i32.
+    if (llvm::isa<FloatType>(srcType)) {
+      src = rewriter.create<LLVM::BitcastOp>(loc, llvmSrcIntType, src);
+    }
+
+    if (srcType.getIntOrFloatBitWidth() < 32) {
+      auto llvmVecType = typeConverter->convertType(mlir::VectorType::get(
+          32 / srcType.getIntOrFloatBitWidth(), llvmSrcIntType));
+      Value undefVec = rewriter.create<LLVM::UndefOp>(loc, llvmVecType);
+      src = rewriter.create<LLVM::InsertElementOp>(
+          loc, undefVec, src, createI32Constant(rewriter, loc, 0));
+      src = rewriter.create<LLVM::BitcastOp>(loc, llvmI32Type, src);
+    }
+
+    // This is taken from the following file llvm/lib/Target/AMDGPU/SIDefines.h
+    enum DppCtrl : unsigned {
+      ROW_SHL0 = 0x100,
+      ROW_SHR0 = 0x110,
+      ROW_ROR0 = 0x120,
+      WAVE_SHL1 = 0x130,
+      WAVE_ROL1 = 0x134,
+      WAVE_SHR1 = 0x138,
+      WAVE_ROR1 = 0x13C,
+      ROW_MIRROR = 0x140,
+      ROW_HALF_MIRROR = 0x141,
+      BCAST15 = 0x142,
+      BCAST31 = 0x143,
+    };
+
+    auto kind = DppOp.getKind();
+    auto permArgument = DppOp.getPermArgument();
+    int attrValue = 0;
+
+    switch (kind) {
+
+    case DPPPerm::quad_perm:
+      if (auto quadPermAttr = permArgument->dyn_cast<ArrayAttr>()) {
```

Comment:
All these `if (auto x = dyn_cast<>())`s should be rewritten to, for example, `auto quadPermAttr = cast<ArrayAttr>(*permArgument);`. If that attribute isn't set up correctly by the time the control flow gets here, assertions should be failing loudly.